syntax = "proto3";
package metadata;

service MetadataDatabase {
    rpc GetByKeys (GetByKeysRequest) returns (GetResponse) {}
    rpc GetByQuery (GetByQueryRequest) returns (GetResponse) {}
    rpc GetKeysByQuery (GetByQueryRequest) returns (GetKeysResponse) {}
    rpc Set (SetRequest) returns (SetResponse) {}
    rpc DeleteTraining (DeleteRequest) returns (DeleteResponse) {}
    rpc RegisterTraining (RegisterRequest) returns (RegisterResponse) {}
    rpc GetTrainingInfo (GetTrainingRequest) returns (TrainingResponse) {}
}

message RegisterRequest {
    int32 training_set_size = 1;
    int32 num_workers = 2;
}

message RegisterResponse {
    int32 training_id = 1;
}

message GetTrainingRequest {
    int32 training_id = 1;
}

message TrainingResponse {
    int32 training_set_size = 1;
    int32 num_workers = 2;
}

message GetByKeysRequest {
    int32 training_id = 1;
    repeated string keys = 2;
}

message GetByQueryRequest {
    string query = 1;
}

message GetResponse {
    repeated string keys = 1;
    repeated float scores = 2;
    repeated string data = 3;
    repeated bool seen = 4;
    repeated int32 label = 5;
}

message GetKeysResponse {
    repeated string keys = 1;
}

message SetRequest {
    int32 training_id = 1;
    repeated string keys = 2;
    repeated float scores = 3;
    repeated bool seen = 4;
    repeated int32 label = 5;
    repeated string data = 6;
}

message SetResponse {
}

message DeleteRequest {
    int32 training_id = 1;
}

message DeleteResponse {
}